feat(mcp): Add session state-based JWT token propagation for MCP tools

timof1308 opened the pull request:
**Please ensure you have read the [contribution guide](https://github.com/google/adk-python/blob/main/CONTRIBUTING.md) before creating a pull request.**

### Link to Issue or Description of Change

**1. Main Tracking Issue:**
This PR is a core component of the main MCP support effort tracked in #3449

**2. Issues Closed/Addressed:**
- Closes: #1402
- Addresses: #805, #2531, #2759, #2168, #5103

**3. Related Open Issues:**
- #5112 — "Secret" session state scope (ephemeral/non-persisted state)
- #5012 — Pass custom HTTP headers to ADK agents in Vertex Agent Engine

### Problem

Currently, MCP tools in ADK lack a secure, standardized way to propagate per-user authentication tokens (like JWTs) from the client to the MCP server. Developers often have to hardcode credentials or modify core FastAPI endpoints to pass these headers, which is not scalable or secure for multi-user environments. Additionally, storing short-lived, sensitive tokens in the persistent `session.state` is a security risk as they may be logged or stored in the database.

### Solution

A mechanism to propagate ephemeral state (`request_state`) from the `RunAgentRequest` through to the `InvocationContext`, plus authentication header support for `McpToolset` at three levels of flexibility:

- **`credential_key`**: Convenience parameter that reads a token from session state and sends it as `Authorization: Bearer <token>`. Aligns with `credential_key` on other ADK toolsets (e.g., `OpenAPIToolset`)
- **`state_header_mapping` / `state_header_format`**: Declarative config for mapping arbitrary state keys to HTTP headers with format templates
- **`header_provider`**: Full control via a callable for dynamic/computed headers

All three levels resolve to the same `HeaderProvider` type internally — no separate code paths. They can be freely combined.

Key changes:
- `request_state`: Added to `InvocationContext` for ephemeral data that overrides `session.state` but is not persisted
- `ReadonlyContext.state`: Merges `request_state` over `session.state` via `ChainMap` (ephemeral takes precedence)
- `_internal.py`: RFC 7230-compliant header validation and sanitization utilities
- Boundary sanitization in both `McpToolset._execute_with_session` and `McpTool._run_async_impl`
- `create_session_state_header_provider`: Utility to generate header provider functions
- Config validation on `McpToolsetConfig` (rejects invalid header names, orphaned format strings, CRLF injection)

### Review Fixes (latest commit)

Addressed code review findings:
- Consolidated header sanitization to boundary methods only (`_execute_with_session`, `_run_async_impl`), removing redundant intermediate sanitization
- Replaced verbose `_DANGEROUS_CHARS` set literal with concise `frozenset` comprehension
- Removed redundant CRLF strip in `create_session_state_header_provider` (already handled by `sanitize_header_value`)
- Switched to lazy %-formatting in logging calls
- Clarified docstrings for `Runner.run()` Optional `new_message` and `McpToolset.from_config()` header_provider limitation

### Testing Plan

- [x] I have added or updated unit tests for my change.
- [x] All unit tests pass locally.

**Unit Tests:**
- `tests/unittests/tools/mcp_tool/test_jwt_token_propagation.py` (37 tests): Header generation, `credential_key`, `state_header_mapping`, `request_state` precedence, RFC 7230 validation, config parsing, CRLF injection protection, combined provider duplicate warnings.
- `tests/unittests/agents/test_readonly_context_state.py` (3 tests): `ReadonlyContext.state` merges ephemeral and persistent state with correct precedence and immutability.
- `tests/unittests/agents/test_readonly_context.py`: No regressions.
- `tests/unittests/tools/mcp_tool/test_mcp_toolset.py`: No regressions in existing toolset functionality.

**Manual End-to-End (E2E) Tests:**

Verified against a local FastMCP server with a mock LLM agent:
- Sent `run_agent` with `request_state={"jwt_token": "test-token-123"}`
- MCP server received header `Authorization: Bearer test-token-123`
- Agent successfully retrieved the echoed value, confirming full propagation

### Checklist

- [x] I have read the [CONTRIBUTING.md](https://github.com/google/adk-python/blob/main/CONTRIBUTING.md) document.
- [x] I have performed a self-review of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] New and existing unit tests pass locally with my changes.
- [x] I have manually tested my changes end-to-end.
- [x] Any dependent changes have been merged and published in downstream modules.

### Usage Examples

#### Option 1: `credential_key` (simplest — Bearer token from state)

**YAML:**
```yaml
tools:
  - name: google.adk.tools.mcp_tool.McpToolset
    args:
      streamable_http_connection_params:
        url: http://api.example.com/mcp
      credential_key: jwt_token
```

**Python:**
```python
from google.adk.tools.mcp_tool import McpToolset

toolset = McpToolset(
    connection_params=StreamableHTTPConnectionParams(
        url='http://api.example.com/mcp'
    ),
    credential_key="jwt_token",
)
```

#### Option 2: `state_header_mapping` (custom headers and formats)

**YAML:**
```yaml
tools:
  - name: google.adk.tools.mcp_tool.McpToolset
    args:
      streamable_http_connection_params:
        url: http://api.example.com/mcp
      state_header_mapping:
        jwt_token: Authorization
        tenant_id: X-Tenant-ID
      state_header_format:
        Authorization: "Bearer {value}"
```

**Python:**
```python
from google.adk.tools.mcp_tool import create_session_state_header_provider, McpToolset

toolset = McpToolset(
    connection_params=StreamableHTTPConnectionParams(
        url='http://api.example.com/mcp'
    ),
    header_provider=create_session_state_header_provider(
        state_key="jwt_token",
        header_name="Authorization",
        header_format="Bearer {value}"
    )
)
```

#### Client-side: Passing the token

```python
# Ephemeral (recommended) — token NOT persisted to session history
requests.post(
    "/run",
    json={
        "app_name": "my_app",
        "user_id": "user123",
        "session_id": "session_id",
        "new_message": {"parts": [{"text": "query"}]},
        "request_state": {
            "jwt_token": "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9..."
        }
    }
)

# Persistent — token saved to session history
requests.post(
    "/run",
    json={
        "app_name": "my_app",
        "user_id": "user123",
        "session_id": "session_id",
        "new_message": {"parts": [{"text": "query"}]},
        "state_delta": {
            "jwt_token": "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9..."
        }
    }
)
```

ryanaiagent commented:
Hi @timof1308 , Thank you for your work on this pull request. We appreciate the effort you've invested. 
Before we can proceed with the review can you please fix the failing unit tests and lint errors

eranc-google commented:
+1

timof1308 commented:
hi @ryanaiagent, rebased with upstream main and reapplied `./autoformat.sh` ✅

timof1308 commented:
hi @ryanaiagent, I've pushed a fix for the unit tests

The failures in `test_readonly_context.py` and `test_mcp_instruction_provider.py` were regressions caused by my changes (missing request_state in the test mocks), which are now resolved and should pass now

however I noticed other failures in the logs (related to pubsub and litellm), but those appear to be present in the upstream adk-python repo

joar commented on src/google/adk/tools/mcp_tool/_internal.py:
```diff
@@ -0,0 +1,317 @@
+    header_format: str = "Bearer {value}",
+    default_value: str = None,
+    strict: bool = False,
+):
```
I think this is a duplicate

timof1308 replied:
@joar Good catch! You're right - this was duplicated

ryanaiagent commented:
Hi @timof1308 , Thank you for your patience here and fixing the unit tests. This PR has merge conflicts that require changes from your end. Could you please rebase your branch with the latest main branch to address these? Once this is complete, please let us know so we can proceed with the review.thanks for fixing them.

timof1308 commented:
hi @ryanaiagent 
I've rebased the branch with the latest upstream/main and resolved the merge conflicts; ready for review and waiting for your feedback

timof1308 committed:
feat: add ephemeral request_state for secure token handling

Add a non-persisted request_state dict to InvocationContext that is
threaded through Runner.run_async and the AdkWebServer. ReadonlyContext.state
now returns a ChainMap merging request_state over session.state, so
ephemeral keys (e.g. tokens) take precedence without being persisted.

timof1308 committed:
feat(mcp): add state-to-header propagation with RFC 7230 validation

Introduce create_session_state_header_provider and create_combined_header_provider
for extracting session state values into HTTP headers with automatic sanitization.
Add credential_key shorthand for Bearer token propagation, state_header_mapping
config for arbitrary state-to-header mappings, and strict mode for type validation.
Header names and values are validated per RFC 7230 to prevent injection attacks.

timof1308 committed:
test: add tests for request_state and header propagation

Add comprehensive tests for request_state merging in ReadonlyContext,
header provider creation, state_header_mapping config, credential_key
shorthand, RFC 7230 header validation, CRLF injection prevention, and
strict mode. Update existing mocks to include request_state.

timof1308 committed:
fix: remove duplicate credential_key shorthand after upstream merge

Upstream merged credential_key for AuthConfig in McpToolset (commit 282db876).
Remove the PR's credential_key Bearer token shorthand to avoid the duplicate
parameter conflict. The same use case is covered by state_header_mapping
with state_header_format.

timof1308 committed:
fix: add missing Apache 2.0 license header to test_readonly_context_state.py

The header-check CI was failing because this new test file was missing
the required copyright/license header.

timof1308 committed:
fix(mcp): harden header handling and add config validation

- Add CRLF/TAB to _DANGEROUS_CHARS and sanitize auth headers
- Standardize header merge order (header_provider takes precedence)
- Add model validator for state_header_mapping/state_header_format
- Warn on duplicate header names in combined providers
- Add request_state/state_delta/invocation_id to sync Runner.run()
- Use TYPE_CHECKING guard in types.py

timof1308 committed:
fix(mcp): sanitize headers in McpTool for consistency with McpToolset

timof1308 committed:
fix(mcp): consolidate header sanitization and clean up review findings

- Move sanitization from _get_auth_headers to _execute_with_session boundary
- Replace verbose _DANGEROUS_CHARS literal with frozenset comprehension
- Remove redundant CRLF strip already handled by sanitize_header_value
- Use lazy %-formatting instead of f-strings in logging
- Clarify docstrings for run() Optional new_message and from_config()
